### PR TITLE
126 I2C IR sensors

### PR DESCRIPTION
**Pull request by aninternetof**
This moves us from the simple GPIO distance sensors to our new I2C ones.
At start-up we:
* Get the I2C port and address from rovercode-web
When a websocket connects:
* Instantiate two instances of the "driver" (helper python module), which initializes the sensor.
* Start the polling thread

I chose to use smbus directly. The Adafruit abstraction library does support I2C, but it doesn't really provide any more abstraction.

Future work:
* At the moment, only the binary value is fed up to Mission Control. In the future, it would be cool to provide a real distance value.
* I chose to make the sensitivity settings rover environment variables and not web configurable. I hope that we eventually come up with an algorithm to make the sensitivity settings unnecessary, so I don't think it's worth it to plumb it up to rovercode-web.

Watch them in action here: https://youtu.be/2SxMV4AEa5s

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,3 +0,0 @@
-{
```
Comment (aninternetof) [inline .vscode/settings.json]: Sorry for having committed this in the first place.

File: setup.sh [hunk 2]
```diff
@@ -15,7 +15,7 @@ fi
 
 apt-get install -y python python-dev python-pip python-smbus
 pip install virtualenv && \
-virtualenv env && \
+virtualenv --system-site-packages env && \
```
Comment (aninternetof) [inline setup.sh]: This ensures that python-smbus is available inside the virtualenv.

File: www/drivers/VCNL4010.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        except (IOError, NameError):
+            print "VCNL4010 was unable to connect to I2C bus {}. " \
+                  "Mocking out the bus".format(i2c_port)
+            self.bus = MockSmbus()
+
+        self.i2c_addr = i2c_addr
+
+        # command register, 0x80
+        # 0xFF(255) Enable ALS and proximity measurement, LP oscillator
+        self.bus.write_byte_data(i2c_addr, 0x80, 0xFF)
```
Comment (cabarnes) [inline www/drivers/VCNL4010.py]: What does the real SMBus do if there is a nak? Do you need to handle an exception or error return value?

File: www/drivers/VCNL4010.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        # LED current register, 0x83
+        # variable led_current*10 mA
+        self.bus.write_byte_data(i2c_addr, 0x83, led_current)
+
+        # ambient light register, 0x84(132)^M
+        # 0x9D Continuos conversion mode, ALS rate 2 samples/sec
+        self.bus.write_byte_data(i2c_addr, 0x84, 0x9D)
+
+    def get_values(self):
```
Comment (cabarnes) [inline www/drivers/VCNL4010.py]: I would suggest adding a test for this method

File: www/drivers/VCNL4010.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        # Read data back from 0x85, 4 bytes
+        # luminance MSB, luminance LSB, Proximity MSB, Proximity LSB
+        data = self.bus.read_i2c_block_data(self.i2c_addr, 0x85, 4)
+
+        # Convert the data
+        luminance = data[0] * 256 + data[1]  # lux
+        proximity = data[2] * 256 + data[3]  # bigger means closer
+        return proximity, luminance
+
+    def is_high(self):
```
Comment (cabarnes) [inline www/drivers/VCNL4010.py]: I would suggest adding a test for this method

File: setup.sh (hunk 2 above)
Comment (aninternetof) [reply setup.sh]: Nope, removed.

File: www/app.py [hunk 6]
```diff
@@ -101,10 +103,17 @@ def read(self):
         result = requests.get(ROVERCODE_WEB_REG_URL+'?client_id='+self.client_id, headers=self.auth_header)
         try:
             info = json.loads(result.text)[0]
+            print info
```
[the remark this answers is not in the document]
Comment (aninternetof) [reply www/app.py]: Nope, removed.

File: www/drivers/VCNL4010.py (hunk 5 above)
Comment (aninternetof) [reply www/drivers/VCNL4010.py]: Done.

File: www/drivers/VCNL4010.py (hunk 4 above)
Comment (aninternetof) [reply www/drivers/VCNL4010.py]: Done.

File: www/drivers/VCNL4010.py (hunk 3 above)
Comment (aninternetof) [reply www/drivers/VCNL4010.py]: Good call. They throw IOError exceptions. I added a handler that just prints an error message and continues. Eventually, we'll come up with a red light or something for the rover to show that something's not working.

**Comment by aninternetof** [comment]
The build is failing because the Travis-CI environment can't find the module `smbus` when it tries to mock it during the tests. This is odd, because we do `apt-get install python-smbus` during the Travis set up, and that should be all that is needed.

I'm headed out of town until next week, so sadly I won't get to work on it until then. If anybody else wants to dig into it, that'd be great!

**Comment by aninternetof** [comment]
Testing is fixed with @cabarnes change!

**Pull request merged**